RES-367: New online clustering algorithm for TM states

marionleborgne opened the pull request:
To run: `python run_experiments_v2.py`

cc @ywcui1990

marionleborgne commented:
Sample output:
```
Exp #0
[DEBUG] New cluster 0 created!
[DEBUG] Cluster 0 permanently added
[DEBUG] New cluster 1 created!
--> index: 700
--> numClusters: 1
--> actualLabel: 2.0
--> predictedClusterLabel: None
--> confidence: -3
--> anomalyScore: 0.9
--> frequencies of actual categories in cluster 0
[
    {
        "actualCategory": 1, 
        "numberOfPoints": 14
    }
]
--------------------------------
[DEBUG] Cluster 1 permanently added
[DEBUG] New cluster 2 created!
--> index: 800
--> numClusters: 2
--> actualLabel: 3.0
--> predictedClusterLabel: 1
--> confidence: 0.516075622196
--> anomalyScore: 0.0
--> frequencies of actual categories in cluster 0
[
    {
        "actualCategory": 1, 
        "numberOfPoints": 14
    }
]
--> frequencies of actual categories in cluster 1
[
    {
        "actualCategory": 2, 
        "numberOfPoints": 59
    }
]

...
```

I still need to work on (this will be a separate PR):
* Better merging mechanism
* Performance: Doesn't scale well with the number of clusters at the moment. And I want to add a compressed representation of the the points in each cluster (instead of keeping all the points in memory). There is also a need for a forgetting mechanism if the points are too old. TBD.

ywcui1990 commented on htmresearch/frameworks/classification/utils/sensor_data.py:
```diff
@@ -110,16 +110,17 @@ def cleanTitle(expResult):
+  #       cleanTitle.append(word)
+  # 
+  # title = ' + '.join(cleanTitle)
+  title = ''
```
This function should be cleaned up

marionleborgne replied:
Thanks for catching that. I committed this by error. I fixed it.

ywcui1990 commented on htmresearch/frameworks/classification/utils/traces.py:
```diff
@@ -165,6 +165,9 @@ def loadTraces(fileName):
+                            'anomalyScore',
+                            'clusteringConfidence',
+                            'predictedClusterLabel',
                             'predictedCategoryTrace']:
```
Are "predictedCategory" and "predictedCategoryTrace" duplicate headers?

marionleborgne replied:
Correct. Another good catch. I'll remove it `predictedCategoryTrace `, it's unused.

ywcui1990 commented on htmresearch/frameworks/clustering/distances.py:
```diff
@@ -70,3 +72,24 @@ def clusterDistDirected(c1, c2):
+  if len(clusters) > 0:
+    for k in range(len(clusters) - 1):
+      c1 = clusters[k]
+      c2 = clusters[k + 1]
```
Why compare the k and the k+1th existing clusters here?

marionleborgne replied:
updated the logic. thanks.

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+    return sorted(dists)
+
+
+  def addCluster(self, cluster):
```
Should there be a removeCluster function?

marionleborgne replied:
Right now there is no need for a remove cluster function. We only merge clusters.

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+
+    clusterDistPairs = self.findClosestClusters(cluster)
+    if len(clusterDistPairs) > 0:
+      notMerged = True
```
The logic can be simplified a bit here. Move notMerged=True outside the if statement, eliminate the else condition. If notMerged, addCluster. No need for two addCluster below.

marionleborgne replied:
good point. updated.

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+          notMerged = False
+      if notMerged:
+        self.addCluster(cluster)
+    else:
```
Eliminate this else condition, and move the if notMerged outside of the first if condition.

marionleborgne replied:
address along with the previous comment. thanks.

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+    self.mergeThreshold = mergeThreshold
+
+
+  def findClosestClusters(self, cluster):
```
The name is confusing as the function returns more than the closest clusters. How about computeClusterDistances?

marionleborgne replied:
agreed, it's clearer. I updated the name.

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+    """
+    clusterDistPairs = self.findClosestClusters(self.newCluster)
+    if len(clusterDistPairs) > 0:
+      distToCluster, predictedCluster = clusterDistPairs[0]
```
What if the newCluster has zero (or very low) distance to more than one clusters?

marionleborgne replied:
I don't think this scenario can happen. If the newCluster has 0 (or very low) distance to more than one clusters, doesn't this mean that the distance between the two clusters is 0? therefore one of these clusters should have been merged into the other one instead of being added as a new cluster.

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+
+    # The data is anomalous
+    if self.anomalousThreshold <= anomalyScore:
+      if self.newCluster.size() > 0:
```
Introduce a threshold here? To make sure that the new clusters are not too small

marionleborgne replied:
I added a threshold

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+      confidence = -3
+
+    # If the data is unstable, so do nothing
+    elif self.stableThreshold <= anomalyScore < self.anomalousThreshold:
```
This could be simplified as `anomalyScore >= self.stableThreshold`

marionleborgne replied:
updated

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -0,0 +1,213 @@
+
+  def merge(self, cluster):
+    for point in cluster.getPoints():
+      self.add(point)
```
check whether point exist in cluster already before merging

marionleborgne replied:
I'd rather implement what we talked about instead: add new point _only_ if you don't find any point in the cluster where `d(newPoint, point) < someThreshold`

ywcui1990 commented:
@marionleborgne done with review. I left some comments.

marionleborgne commented:
@ywcui1990 I think I addressed all your comments. I also added the logic to only add a new point to cluster only of the distance to the closest point of the cluster is high enough. This is reducing the number of points in each clusters a bit (depending on how you set the `similarityThreshold` value)

ywcui1990 commented:
The change looks good to me. Does the similarityThreshold solve the problem of increasing clustering size over learning?

marionleborgne commented:
Yes, `similarityThrehsold` does help with the increasing size of each cluster. If `similarityThrehsold=0` then the sum of all cluster sizes is ~310 points.  If `similarityThrehsold=0.01` the sum of all cluster sizes is 250 points.

marionleborgne commented:
I have added additional changes:
* Add cluster stats (lastUpdated, timeCreated, numberPointsClustered)
* Add pruning and allow to turn on or off.
* Break ties with lastUpdated.

If you tweak `merginThrehsold` correctly you get good results:
<img width="1218" alt="clustering_signal binary_mergingthreshold 0 3" src="https://cloud.githubusercontent.com/assets/2488251/18652617/b1be811a-7e88-11e6-98b3-8f9ca64e08c6.png">


```
[DEBUG] -> frequencies of actual categories in cluster 8.
[DEBUG] -> cluster info: {'_timeCreated': 7, '_label': 1, 'numPoints': 9, '_numPointsClustered': 9, '_lastUpdated': 46, '_id': 8}
[DEBUG] * actualCategory: 1
[DEBUG] * numberPoints: 5
[DEBUG] 
[DEBUG] * actualCategory: 2
[DEBUG] * numberPoints: 4
[DEBUG] 
[DEBUG] -> frequencies of actual categories in cluster 12.
[DEBUG] -> cluster info: {'_timeCreated': 16, '_label': 1, 'numPoints': 58, '_numPointsClustered': 58, '_lastUpdated': 979, '_id': 12}
[DEBUG] * actualCategory: 1
[DEBUG] * numberPoints: 58
[DEBUG] 
[DEBUG] -> frequencies of actual categories in cluster 59.
[DEBUG] -> cluster info: {'_timeCreated': 91, '_label': 2, 'numPoints': 6, '_numPointsClustered': 6, '_lastUpdated': 649, '_id': 59}
[DEBUG] * actualCategory: 2
[DEBUG] * numberPoints: 6
[DEBUG] 
[DEBUG] -> frequencies of actual categories in cluster 24.
[DEBUG] -> cluster info: {'_timeCreated': 30, '_label': 2, 'numPoints': 47, '_numPointsClustered': 47, '_lastUpdated': 647, '_id': 24}
[DEBUG] * actualCategory: 2
[DEBUG] * numberPoints: 47
[DEBUG] 
[DEBUG] -> frequencies of actual categories in cluster 58.
[DEBUG] -> cluster info: {'_timeCreated': 89, '_label': 2, 'numPoints': 133, '_numPointsClustered': 133, '_lastUpdated': 949, '_id': 58}
[DEBUG] * actualCategory: 2
[DEBUG] * numberPoints: 133
```

Accuracy:
```
[DEBUG] -> recordNumber: 950
[DEBUG] -> sensorValue: 3.04837
[DEBUG] -> actualCategory: 0.0
[DEBUG] -> anomalyScore: 0.85
[DEBUG] -> classificationInference: 1
[DEBUG] -> classificationAccuracy: 0.263886372685 / 1
[DEBUG] -> clusteringInference: None
[DEBUG] -> predictedClusterId: None
[DEBUG] -> clusteringAccuracy: 0.970584237421 / 1
[DEBUG] -> clusterHomogeneity: 98.406374502 / 100
[DEBUG] -> clusteringConfidence: -3
[DEBUG] -> numClusters: 5
```

marionleborgne commented:
@ywcui1990 done with my changes

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -90,10 +103,13 @@ def __init__(self,
 
     # Clusters
-    self._newCluster = Cluster(fistClusterId)
+    self._numPointsClustered = 0
```
The name is a little confusing, as we are not adding every single input to the clusters. How about iteration or steps?

marionleborgne replied:
👍 updated

ywcui1990 commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -142,31 +148,82 @@ def addCluster(self, cluster):
+
+  def _pruneClusters(self):
+    clusters = self.getClusters()
+    for i in range(len(clusters)):
```
I have a question on the logic here. Say that you have 3 clusters, c0 and c1 have many shared points, c1 and c2 have many shared points, but c0 and c2 doesn't have that many shared points (above threshold).

My question is what will happen in this case. Does the result depend on the order of the clusters?

marionleborgne replied:
As per our discussion, since we don't need and don't use pruning at the moment, I marked it as not implemented. If we do use pruning in the future, we should iterate through all permutations of remaining clusters to make sure we are not missing a cluster that should actually be merged.

marionleborgne commented on htmresearch/frameworks/clustering/distances.py:
```diff
@@ -96,3 +96,25 @@ def interClusterDistances(clusters, newCluster):
+    dists.append((d, c))
+
+  clusterDists = sorted(dists, key=lambda x: (x[0], x[1]._lastUpdated))
+  # print 'SORTED by distance:'
```
👍

marionleborgne commented on htmresearch/frameworks/clustering/sdr_clustering.py:
```diff
@@ -142,31 +148,82 @@ def addCluster(self, cluster):
+
+    clusterDistPairs = computeClusterDistances(cluster, clusters)
+    clusterMerged = False
+    if len(clusterDistPairs) > 0:
```
Well, yes, if  `clusterDistPairs ` is an empty list (i.e. there are no clusters yet (just the `newCluster`) then the block below the if statement makes no sense (and there is nothing to merge)
